src/main/java/echo/tasklist/TaskList.java

```diff
@@ -0,0 +1,124 @@
+     * Marks task with given value and returns the task that is marked
+     *
+     * @param value index of the task to be marked in the taskList.
+     * @return the task being marked.
+     */
+    public Task markAndGetTask(String value) {
+        int index = Integer.parseInt(value) - 1;
+        Task task = this.getTask(index);
+        task.mark();
+        return task;
+    }
+
+    /**
+     * Unmarks task with given value and returns the task that is unmarked
+     *
+     * @param value index of the task to be unmarked in the taskList.
+     * @return the task being unmarked.
+     */
+    public Task unmarkAndGetTask(String value) {
+        int index = Integer.parseInt(value) - 1;
+        Task task = this.getTask(index);
+        task.unmark();
+        return task;
+    }
+
+    /**
+     * Returns the Task at a given index of taskList
+     *
+     * @param index of the task in taskList
+     * @return Task
+     * @throws EchoException If index is greater than the total number of tasks in taskList
+     */
+    public Task getTask(int index) {
+        int largestIndex = this.sizeOfTaskList() - 1;
+        if (index > largestIndex) {
+            throw new EchoException("There is not enough task. "
+                    + "\nPlease add more task or change another index.");
+        }
+        return this.taskList.get(index);
+    }
+
+    /**
+     * Deletes Task at a given index in taskList
+     *
+     *
+     * @throws EchoException If index is greater than the total number of tasks in taskList
+     */
+    public Task getTaskAndDelete(String taskDescription) throws EchoException {
+        int index = Integer.parseInt(taskDescription) - 1;
+        int largestIndex = this.sizeOfTaskList() - 1;
+
+        if (index > largestIndex) {
+            throw new EchoException("There is not enough task. "
+                    + "\nPlease add more task or change another index.");
+        }
+
+        Task task = this.taskList.get(index);
+
+        this.taskList.remove(index);
+        Collections.rotate(this.taskList.subList(index, this.sizeOfTaskList()), -1);
+        this.taskList.trimToSize();
+        return task;
+    }
+
+    /**
+     * Adds Task into taskList and print a display message
+     *
+     * @param userTask Task provided by user input
+     */
+    public void addTask(Task userTask) {
+        this.taskList.add(userTask);
+    }
+
+    /**
+     * Finds task in taskList with keyword in the description and return it as an ArrayList of task
+     *
+     * @param keyword word to search for in the list of task
+     * @return an arrayList of all the task with description containing the keyword
+     */
+    public ArrayList<Task> findTask(String keyword) {
+        ArrayList<Task> resultArray = new ArrayList<>();
+        for (Task task : this.taskList) {
+            String description = task.getDescription();
+            if (description.contains(keyword)) {
+                resultArray.add(task);
+            }
```

Comment:
Could this logic be extracted out into the `Task` class directly? This could help with the abstraction of the `Task` class and hide its internal implementation details.